File: lib/src/resources/signaling_api_provider.dart
import "dart:async";
import "package:http/http.dart" as http;
import "dart:convert";

import "../../settings.dart";

// Required types
enum SignalingResponse { SUCCESSFULL, NO_DEVICE, NO_DATA }

class SignalingApiProvider {
  Future<List<dynamic>> getUserDevices(String userId) async {
    var response = await http.get("$baseUrlSignaling/user/$userId/devices");
    return jsonDecode(response.body);
  }

  Future<SignalingResponse> sendData(
      String userId, String deviceId, Map<String, dynamic> data) async {
    var response = await http.post(
        "$baseUrlSignaling/user/$userId/device/$deviceId",
        headers: {"Content-Type": "application/json"},
        body: jsonEncode({"data": data}));

    switch (response.statusCode) {
      case 200:
        return SignalingResponse.SUCCESSFULL;
      case 400:
        return SignalingResponse.NO_DATA;
      case 404:
        return SignalingResponse.NO_DEVICE;
      default:
        return SignalingResponse.NO_DEVICE;
    }
  }
}

File: ios/Runner/SignalingApiProvider.swift
//
//  SignalingApiProvider.swift
//  Runner
//
//  Created by Sudharshan on 3/29/19.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

import Foundation
import WebRTC
import Just

class SignalingApiProvider: NSObject {
    var authToken: String?
    
    public override init() {
        super.init()
    }
    
    public convenience init(authToken: String) {
        self.init()
        self.authToken = authToken
    }
    
    public func getUserDevices(userId: String) -> [String]? {
        var deviceList: [String] = []
        let response = Just.get("http://localhost/signal/user/\(userId)/devices",
            headers: ["Authorization": "Bearer \(authToken ?? "0")"])
        
        if(response.ok) {
            do {
                if let jsonResult = try JSONSerialization.jsonObject(with: response.content!, options: []) as? [String] {
                    // convert this to an array of strings
                    for device in jsonResult {
                        deviceList.append(device)
                    }
                    
                    return deviceList
                } else {
                    // Invalid response format
                    return nil
                }
            } catch let error as NSError {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    public func getConversationUsers(conversationId: String) -> [String]? {
        var userList: [String] = []
        let response = Just.get("http://localhost/core/user/conversation/\(conversationId)/member",
            headers: ["Authorization": "Bearer \(authToken ?? "0")"])
        
        if (response.ok) {
            do {
                if let jsonResult = try JSONSerialization.jsonObject(with: response.content!, options: []) as? [Any] {
                    // Need code to convert this to an array of strings
                    for user in jsonResult {
                        if let userObject = user as? [String: String] {
                            guard let userId = userObject["id"] else {
                                // Invalid response format
                                return nil
                            }
                            
                            userList.append(userId)
                        }
                    }
                    
                    return userList
                }
            } catch let error as NSError {
                print(error.localizedDescription)
            }
        }
        return nil
    }
    
    // CHECK FOR WHEN DEVICE IS UNAVAILABLE
    public func postDataToUser(userId: String, deviceId: String, data: String, event: String) {
        Just.post("http://localhost/signal/user/\(userId)/device/\(deviceId)",
            json: ["event": event, "data": data],
            headers: ["Authorization": "Bearer \(authToken ?? "0")"])
    }
}
